std/prelude.glu

```diff
@@ -207,13 +207,13 @@ let category_Function : Category (->) = {
     compose = \f g x -> f (g x)
 }
 
-let make_Category cat =
+let make_Category cat : Category cat -> _ =
```

Comment:
Would it be possible to throw an error, saying that all generic types must be explicitly quantified? So you would need to write:

```
let make_Category cat : forall cat . Category cat -> _ =
```

Reply:
I went with inserting forall's implicitly since that is done anyway if one omits the type declaration.

```f#
let id x = x // works
let id x : a -> a = x // Error: `a` is not defined
let id x : forall a . a -> a = x // Have to add forall
```

It just seems a bit counterproductive that omitting the type entirely lets the type be inferred but typing in the variables causes it to fail... :/ Feel free to convince me otherwise though.

Reply:
https://github.com/purescript/purescript/issues/766#issuecomment-165842979

> I believe explicit scoping right off the bat helps prevent confusion when moving to higher-rank types and such. Hiding System F for a special (if common) case hides a key fact about how types really work.

https://github.com/purescript/purescript/issues/766#issuecomment-167426170

> I'd also suggest that many beginning-to-intermediate Haskell programmers have an inaccurate understanding of what something like `id :: a -> a` means. I'm not sure exactly how to describe it, but it's like they picture the preceding id function as being a function from some unknown type a to that same type.

Reply:
I just really hate ML and Haskell for mysteriously magicking up unbound identifiers 😆

Reply:
I like to view it a bit like lifetime elision in rust. However, since type variables are scoped in gluon I'd say that is a really good argument for forcing `forall`...

On another note, `forall` is not a very good keyword to declare variables unless one already knows about universal quantification... Reminds me of `public static void main` in Java. 

> What does `forall` mean?

Uhh, it just lets you write declare generic paramaters.

> But what does it **mean** ?

**Points to stack of university literature and runs away**

Reply:
It mean's 'for all types `a`, there is a function called `id` defined that goes from `a` to `a`'.

Reply:
Personally I prefer a syntax that makes it look more like a function though. Like:

```
id: ?(a : Type) -> a -> a
```

That way it's clear that it's just another parameter, albeit at the type level.

Reply:
> It mean's 'for all types a, there is a function called id that goes from a to a'.

That is a good explanation but I still feel like that explanation can be difficult to grok from someone coming from Java, Rust etc.

> id: ?(a : Type) -> a -> a

I do like that on a theoretical level, but on a practical level I fear it makes generic functions more cumbersome than non-generic ones

```
// The first function is the better/more general function but it is much shorter to write the second...
// Each additional type parameter makes the difference larger as well
let id x: ?(a : Type) -> a -> a = x
let id x: Int -> Int = x
```

Reply:
Perhaps we could still allow folks to drop the kind annotations. 🤔

Other langs allow folks to bunch the implicits together:

```rust
let compose : ?[a b c] -> (b -> c) -> (a -> b) -> a -> c
```

Then annotate them by name:

```rust
let gf = compose ?[c=Int] ?[a=Float] g f
```

Came across an interesting post on the problems with Agda's implicit arguments btw: https://github.com/bitonic/tog/wiki/Implicit-Arguments - all seems rather complicated :/

Reply:
> `let compose : ?[a b c] -> (b -> c) -> (a -> b) -> a -> c`

That looks really clean actually. It kind of splits the middle between `<A, B, C>/[A, B, C]` declared parameters and having a common syntax for types and values. (`[]` has some precedence as use for generic variables as well)

> Came across an interesting post on the problems with Agda's implicit arguments btw

Will take a look!